dfusion_rust_core/src/models/solution.rs

```diff
@@ -17,12 +17,79 @@ pub struct Solution {
 impl Solution {
     pub fn trivial(num_orders: usize) -> Self {
         Solution {
-            objective_value: Some(U256::zero()),
             prices: vec![0; TOKENS as usize],
             executed_buy_amounts: vec![0; num_orders],
             executed_sell_amounts: vec![0; num_orders],
         }
     }
+
+    /// Returns the token price for token or `None` if the token is invalid.
+    fn price(&self, token: u16) -> Option<u128> {
+        self.prices.get(token as usize).cloned()
+    }
+
+    /// Compute the result of the objective function for the solution. This is
+    /// currently defined as *U - dUT* where *U* is the total utility of the
+    /// trade and *dUT* is the total disreguarded utility of touched orders.
+    /// Note that this does not check the validity of the solution.
+    ///
+    /// # Returns
+    ///
+    /// Returns the objective value for the solution or None if there the orders
+    /// do not match the solution or there was an error in the calculation such
+    /// as overflow or divide by 0.
+    pub fn objective_value(&self, orders: &[Order]) -> Option<U256> {
+        if orders.len() != self.executed_buy_amounts.len()
+            || orders.len() != self.executed_sell_amounts.len()
+        {
+            return None;
+        }
+
+        // note that we don't actually have to do the same checking here (such as
+        // account balances and such) as we do in the smart contract - we only
+        // care if the solution is valid given the provided orders
+
+        let mut total_utility = U256::zero();
+        let mut total_disregarded_utility = U256::zero();
+
+        let mut fee_buy_amount = U256::zero();
+        let mut fee_sell_amount = U256::zero();
+
+        for (i, order) in orders.iter().enumerate() {
+            let buy_price = self.price(order.buy_token)?;
+            let sell_price = self.price(order.sell_token)?;
+            let exec_buy_amount = self.executed_buy_amounts[i];
+            let exec_sell_amount = self.executed_sell_amounts[i];
+
+            if exec_buy_amount == 0 && exec_sell_amount == 0 {
+                // this order was not touched, so skip
+                continue;
+            }
+
+            total_utility = total_utility.checked_add(order.utility(
+                buy_price,
+                exec_buy_amount,
+                exec_sell_amount,
+            )?)?;
+            total_disregarded_utility = total_disregarded_utility.checked_add(
```

Comment:
Do we take the user's balance into account somehow? Disregarded utility should not be positive if the user had no remaining funds.

Reply:
No we currently don't account for balances at all. We could add it for sure, I created #290 for this.